### ISSUE TITLE
Whether to support changing to another algorithm

### ISSUE THREAD

**Issue by pzhabc** [reiniscimurs/DRL-robot-navigation#76]
Hi. Thank you for everything. This is great. I would like to ask if I can change the TD3 algorithm to other algorithms based on your work, such as DQN, DDQN and so on. I don't know much about this. I hope you can spare some time to answer my questions. Thanks again!

**Comment by reiniscimurs** [comment]
Hi
Yes you could change the algorithm to other DRL algorithms and train the network with your own methods. You could specify a different algorithm class to be called in line:
https://github.com/reiniscimurs/DRL-robot-navigation/blob/101ed8eaba77b9f447e3d57cc71518cfedb7f65c/TD3/train_velodyne_td3.py#L261
This would call a different DRL training, assuming that the class methods (such as `train()` and `get_action()`) are implemented.
It should be relatively painless to adapt a DQN or DDQN algorithm to work here.

**Comment by pzhabc** [comment]
I'm sorry to ask you again. Based on your work, I used DQN to implement it. The number of action Spaces of the robot is 5, and the corresponding linear and angular velocities of the 5 actions are shown in the picture below. The neural network outputs the value of the 5 actions, and then the robot chooses one of the actions to execute. However, I encountered some problems during operation, as shown in the picture, after entering the train(), the values of max_value and Q_value are very large, I don't know what caused it, can you give me some help? Looking forward to your reply, thank you!
![bafd4adce2f14e4d5e4d4c1be9e43c31](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/89b200b4-abc4-4eed-958e-a6aaa7a5898c)
![192f48c910bc853d5b1162b76cf8ac71](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/974d6037-cb7e-4645-828f-caa4287472ee)
![ff6d18f50ec23d7f82df2ba0c08041f7](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/1bd4c74d-6efa-4f61-8058-69b19ee851bc)
![902484cc9faaa4b6b78ac8fa7ab8d024](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/a2429ccf-a871-4c36-9bc2-86d595e4c030)
![4e555d86f4cc330a63e6549fecab3438](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/e18cf7d4-4e32-45f1-93c6-fb2b182a340f)

**Comment by pzhabc** [comment]
> 嗨，是的，您可以将算法更改为其他 DRL 算法并使用您自己的方法训练网络。您可以指定要在行中调用的其他算法类：
> 
> https://github.com/reiniscimurs/DRL-robot-navigation/blob/101ed8eaba77b9f447e3d57cc71518cfedb7f65c/TD3/train_velodyne_td3.py#L261
> 
> 
> 这将调用不同的 DRL 训练，假设实现了类方法（例如 和 ）。
> 在这里调整 DQN 或 DDQN 算法应该相对轻松。`train()``get_action()`

I'm sorry to ask you again. Based on your work, I used DQN to implement it. The number of action Spaces of the robot is 5, and the corresponding linear and angular velocities of the 5 actions are shown in the picture below. The neural network outputs the value of the 5 actions, and then the robot chooses one of the actions to execute. However, I encountered some problems during operation, as shown in the picture, after entering the train, the values of max_value and Q_value are very large, I don't know what caused it, can you give me some help? Looking forward to your reply, thank you!
![bafd4adce2f14e4d5e4d4c1be9e43c31](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/9037e99d-d564-4c14-a5d4-9f556ea5e1e1)
![192f48c910bc853d5b1162b76cf8ac71](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/4c5d50b7-db51-40fa-9505-7097a375871d)
![ff6d18f50ec23d7f82df2ba0c08041f7](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/467577a5-187e-4140-90eb-5b516aab165a)
![902484cc9faaa4b6b78ac8fa7ab8d024](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/dce7c8f8-2b22-41d7-a80c-b808ba22eed0)
![4e555d86f4cc330a63e6549fecab3438](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/9bca55a0-0a3b-46b3-9e7b-2bcecc4b2449)

**Comment by reiniscimurs** [comment]
Is the training running alright and at what stage of training do you see these values? How do you come to conclusion that the Q value is too high? If it is a random initialization the Q values would be rather unstable and could be overestimated in the beginning of the training.

**Comment by pzhabc** [comment]
> Is the training running alright and at what stage of training do you see these values? How do you come to conclusion that the Q value is too high? If it is a random initialization the Q values would be rather unstable and could be overestimated in the beginning of the training.

Thanks for your reply! The training did not go well, and it was the first step into the train() function that led to this result.

**Comment by reiniscimurs** [comment]
Can you share the full code?

It seems a bit weird to me that action is reshaped to shape (-1,1) even though it is a 2 dimensional vector in the regular replay buffer. Then this is used in max_q calculation with gather function. If you are using the DQN from pytorch tutorial, then there it uses a 1D vector for action space so there might be a discrepancy there.

**Comment by pzhabc** [comment]
> Can you share the full code?
> 
> It seems a bit weird to me that action is reshaped to shape (-1,1) even though it is a 2 dimensional vector in the regular replay buffer. Then this is used in max_q calculation with gather function. If you are using the DQN from pytorch tutorial, then there it uses a 1D vector for action space so there might be a discrepancy there.

Thanks for your reply! I mainly changed the actions of the robot in the file "train_velodyne_td3.py", but did not change other files. Please look at the picture.
![df8400faa4a4a75806bacef38792f1b7](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/e08dba8f-876d-44c2-a438-c9e8bc470122)
![261e5f2f9b41c2721cb00097c6f9f3e9](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/a416b988-f651-41dc-81ac-85d745d6c19d)
![image](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/d088a293-8031-47b0-b7ac-6f8a0bedb46b)
![100f1f2914cf5ae59288fd2fe8d42d46](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/6097b659-1a55-497f-80dd-e3153bb91328)
![d60af7a21f94f0c85bc4be6f4efca19e](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/1b5f122c-c5ed-4841-bb6e-5f18235f5e33)

**Comment by reiniscimurs** [comment]
Are you implementing the DQN from pytorch (https://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html)?
If not, let me know the source of your implementation.

I do not see any major issues here, but the train function might have some issues here. Not sure if you want to reset the target critic model every second training iteration (line 114). Might be better to reset after the training loop has completed. But generally I would expect a soft update of the target network parameters, which I do not see in this code. Here, the target network just takes over all the network weights as they are on every second weight update. This should make the network quite unstable and learning might fail.

But also did you run it only for one training episode? Do the Q values not reduce over the training? Here we are initializing the neural network with random values so the values being wildly off in the beginning would not be unexpected.

**Comment by pzhabc** [comment]
> Are you implementing the DQN from pytorch (https://pytorch.org/tutorials/intermediate/reinforcement_q_learning.html)? If not, let me know the source of your implementation.
> 
> I do not see any major issues here, but the train function might have some issues here. Not sure if you want to reset the target critic model every second training iteration (line 114). Might be better to reset after the training loop has completed. But generally I would expect a soft update of the target network parameters, which I do not see in this code. Here, the target network just takes over all the network weights as they are on every second weight update. This should make the network quite unstable and learning might fail.
> 
> But also did you run it only for one training episode? Do the Q values not reduce over the training? Here we are initializing the neural network with random values so the values being wildly off in the beginning would not be unexpected.

Thank you very much for your help! I referred to the link you provided. The soft update you mentioned was used. After running several training sessions, you find the results as shown below. Does this mean that the training is effective?
![93a7b2b4f48b11a8cb7ebe09a1941b1e](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/dbb90631-584d-4463-a31b-eab1ede5d206)
![83f3517a545830bed129253567b1ef63](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/0f2574d7-deaf-4ef3-b762-a567804fafe6)
![0620409ef409bf2c279139a53a44ea4c](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/57391001-c4f2-4d57-b23d-1f635aaa99ca)

But I'd like to ask you a few more questions.

1. "Exception sending a multicast message:Network is unreachable" appears on the terminal. Do I need to be connected to the network for this training?

2. The implementation of DQN algorithm I refer to is not the official website, it does not use soft update, but the official use. Here soft updates seem to play a big role. Is it necessary in the algorithm?

3. For some existing algorithms, can I also refer to the official website provided by you for implementation?

4. May I ask what caused the initial increase in the loss function?

Thank you, you have been a great help to me!

**Comment by reiniscimurs** [comment]
1. I am not familiar with this error. Please post the error log.
2.  Here is a good answer to this question: https://ai.stackexchange.com/questions/21485/how-and-when-should-we-update-the-q-target-in-deep-q-learning
While not strictly required, the target network update scheme must make sense.
3. I am not sure what the question is here. Feel free to refer to official websites of implementation if it makes sense.
4. Loss function is the target Q value difference from Q value. Initially you will see smaller target Q values as reward will be low since most likely would not have encountered that many terminal episodes. Then target Q values will be lower and very close to Q values. As such, your loss will increase and then reduce once actual training begins.

It seems to me that some training is happening as average Q value increases. Not sure if you want to cap the max Q value though, but that is up to your design. However, you should be able to visually see how the policy works in the environment. 30 episodes are a rather small amount for any reasonable training though.

**Comment by pzhabc** [comment]
Ok, thank you, what you said is quite right. Before, there was too little training, but with the increase of training, I found a problem. Is it because DQN is not suitable for such a scenario or the code itself has defects and needs to be improved?
![3fadef0498148560976fb25c96d31cfa](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/ab68a39b-bd35-4254-a9b5-348a1b904bc7)

**Comment by reiniscimurs** [comment]
DQN and DDQN should not have any major issues in learning with this code as there exist successful implementations. I do not see any reason why it shouldn't work from theoretical standpoint. Most likely there is some discrepancy in the code itself.

**Comment by pzhabc** [comment]
Okay, thank you. I recently in learning reinforcement learning in this website "https://spinningup.openai.com/en/latest/spinningup/rl_intro3.html" saw such a content: In Policy Learning, we should care about average returns. The loss function means nothing. I wonder if the same is true in value learning?

**Comment by pzhabc** [comment]
> DQN and DDQN should not have any major issues in learning with this code as there exist successful implementations. I do not see any reason why it shouldn't work from theoretical standpoint. Most likely there is some discrepancy in the code itself.

Okay, thank you. I recently in learning reinforcement learning in this website "https://spinningup.openai.com/en/latest/spinningup/rl_intro3.html" saw such a content: In Policy Learning, we should care about average returns. The loss function means nothing. I wonder if the same is true in value learning?
![image](https://github.com/reiniscimurs/DRL-robot-navigation/assets/147294035/601d5cba-8c74-4e72-8b21-d7e32f0e24a1)

**Comment by reiniscimurs** [comment]
What they mean here is that in typical imitation learning tasks you would have ground truth and your calculated observation. Loss is expressed as the difference between ground truth and the observation. Therefore, the lower the loss, the closer your model is to ground truth so that is your target for optimization.

Here, though, there is no ground truth, there is only a reward function that you are trying to optimize for. The loss is just a direction for your optimization in which to optimize your parameters, as you are comparing the target Q to the network Q value, both of which are to be learned.

It does not matter if it is DQN or Policy gradient. What you learn in the end is to estimate the value of your state-action pair. The question for the network here is "if I take this action, what will be my current + future reward?". The more correct your Q value estimation is, the better the policy. Both methods solve this problem.

You could look into the implementation of a similar method with D3QN in: https://arxiv.org/pdf/1706.09829.pdf
You could take inspiration from their open source code.
